### PR TITLE
Real-time Chat

### PR DESCRIPTION
**Pull request by paglias**
WIP
### Implementation
- [ ] Tweaks to interface and notifications on the client
- [ ] What to do with flags
- [ ] Spam blocker
- [ ] Limit concurrent messages
- [ ] Security when booting an user from a group (we must be sure it's been disconnected)
- [ ] Batch updates to `user.newMessages.{groupId}`?
- [ ] is `user.lastMessageSeen` still necessary?
### Technical steps
- [x] Add pusher library to the server 
- [x] Limit to private parties / guilds for now
- [x] Add channel authentication
- [ ] Authenticate multiple channels with 1 HTTP call (only needed for guilds)
- [x] Make sure the user has access to the channels when authenticating
- [x] Add real-time updates to the client
- [ ] Implement party cap
- [ ] Mark a message as seen when the user is on the party page and a new message comes in
- [ ] Implement tweaks to client UI
- [x] use only presence channel for parties
- [x] remove user when is booted or leaves group
- [x] Remove old Firebase code if we definitely decide to not use it anymore
- [x] Tests

### CODE REVIEW

File: website/server/controllers/api-v3/auth.js [hunk 1]
```diff
@@ -282,6 +283,55 @@ api.loginSocial = {
...
 
+/*
+ * @apiIgnore
+ * @api {post} /api/v3/user/auth/pusher Pusher.com authentication
+ * @apiDescription Authentication for Pusher.com private and presence channels, private route
+ * @apiVersion 3.0.0
+ * @apiName UserAuthPusher
+ * @apiGroup User
+ *
+ * @apiParam {String} socker_id Body parameter
```
Comment (khipkin) [inline website/server/controllers/api-v3/auth.js]: socket_id

Comment (paglias) [inline website/server/controllers/api-v3/auth.js]: fixed

File: website/server/controllers/api-v3/auth.js [hunk 2]
```diff
@@ -312,10 +314,30 @@ api.pusherAuth = {
...
+      throw new BadRequest('Invalid Pusher channel type.');
+    }
+
+    if (resourceType !== 'group') { // only groups are supported
+      throw new BadRequest('Invalid Pusher resource type.');
+    }
+
+    resourceId = resourceId.join('-'); // the split at the beginning had split resourceId too
+    if (!validator.isUUID(resourceId)) {
+      throw new BadRequest('Invalid Pusher resource id, must be a UUID.');
```
Comment (paglias) [inline website/server/controllers/api-v3/auth.js:327]: @crookedneighbor do you think these strings should be localized?

Comment (crookedneighbor) [inline website/server/controllers/api-v3/auth.js:327]: These all look like implementation details, so as long as the API is private, I think it's fine to not localize them. A user should never see these errors.

File: test/api/v3/integration/user/auth/POST-user_auth_pusher.test.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  it('returns an error if the passed resource id doesn\'t match the user\'s party', async () => {
+    await expect(user.post(endpoint, {
+      channel_name: `private-group-${generateUUID()}`,
+      socket_id: '123',
+    })).to.eventually.be.rejected.and.eql({
+      code: 404,
+      error: 'NotFound',
+      message: 'Resource id must be the user\'s party.',
+    });
+  });
```
Comment (khipkin) [inline test/api/v3/integration/user/auth/POST-user_auth_pusher.test.js:101]: success case(s)?

Comment (paglias) [inline test/api/v3/integration/user/auth/POST-user_auth_pusher.test.js:101]: The problem is that these are integration tests and we cannot stub the return value of `pusher.authenticate`

File: bower.json [hunk 4]
```diff
@@ -42,6 +42,8 @@
     "swagger-ui": "wordnik/swagger-ui#v2.0.24",
     "smart-app-banner": "78ef9c0679723b25be1a0ae04f7b4aef7cbced4f",
     "habitica-markdown": "1.2.1"
```
Comment (khipkin) [inline bower.json]: Missing comma at EOL

Comment (paglias) [inline bower.json]: good catch! it was why the build was failing

**Comment by khipkin** [comment]
Can probably close #2466 when you finish this and merge it in. :smiley_cat:

**Comment by paglias** [comment]
Going to merge it as it's, will implement the UI separately

**Pull request merged**